File: cli-python/import_contracts.py
#!/usr/bin/env python3
"""
Combined script to setup pipeline, create index, and ingest PDFs.
Runs all steps in sequence for easy deployment.
"""

import sys
import argparse
import requests
import base64
import json
from pathlib import Path

try:
    import yaml
except ImportError:
    yaml = None

SCRIPT_DIR = Path(__file__).resolve().parent
DEFAULT_CONFIG_PATH = SCRIPT_DIR.parent / "config" / "elasticsearch.yml"
DEFAULT_MAPPING_PATH = SCRIPT_DIR.parent / "config" / "mappings-contracts.json"

# Default values
ES_INDEX = 'contracts'
DEFAULT_INFERENCE_ENDPOINT = '.elser-2-elastic'


def load_yaml(path: Path) -> dict:
    """Load YAML configuration file."""
    if not path.exists():
        raise FileNotFoundError(f"Config file not found: {path}")

    if yaml is None:
        raise RuntimeError(
            "PyYAML is required to load configuration files. Install with 'pip install PyYAML'."
        )

    with path.open("r", encoding="utf-8") as handle:
        data = yaml.safe_load(handle) or {}

    if not isinstance(data, dict):
        raise ValueError(f"Configuration must be a mapping (found {type(data).__name__})")
    return data


def load_json(path: Path) -> dict:
    """Load JSON configuration file."""
    if not path.exists():
        raise FileNotFoundError(f"JSON file not found: {path}")

    with path.open("r", encoding="utf-8") as handle:
        data = json.load(handle)

    if not isinstance(data, dict):
        raise ValueError(f"JSON file must define a JSON object (found {type(data).__name__})")
    return data


def build_auth_header(config: dict) -> str:
    """Build authorization header from config (prefer api_key over user/password)."""
    api_key = config.get("api_key", "").strip()
    if api_key:
        return f"ApiKey {api_key}"
    
    user = config.get("user", "").strip()
    password = config.get("password", "").strip()
    if user and password:
        token = f"{user}:{password}"
        encoded = base64.b64encode(token.encode("utf-8")).decode("ascii")
        return f"Basic {encoded}"
    
    raise ValueError("Config must include either 'api_key' or both 'user' and 'password'")


def load_config(config_path: Path = None) -> tuple:
    """Load Elasticsearch configuration and return endpoint and headers."""
    if config_path is None:
        config_path = DEFAULT_CONFIG_PATH
    
    config = load_yaml(config_path)
    
    endpoint = config.get("endpoint", "").strip()
    if not endpoint:
        raise ValueError("Config must include an 'endpoint'")
    
    auth_header = build_auth_header(config)
    headers = {
        "Authorization": auth_header,
        "Content-Type": "application/json"
    }
    
    # Add any custom headers from config
    custom_headers = config.get("headers", {})
    if isinstance(custom_headers, dict):
        headers.update({str(k): str(v) for k, v in custom_headers.items()})
    
    return endpoint, headers


# Global variables (will be set in main)
ES_ENDPOINT = None
headers = None
INFERENCE_ENDPOINT = DEFAULT_INFERENCE_ENDPOINT  # Default, will be auto-detected if not found


def check_elasticsearch():
    """Check if Elasticsearch is reachable."""
    try:
        response = requests.get(ES_ENDPOINT, headers=headers, timeout=5)
        if response.status_code == 200:
            info = response.json()
            print(f"Cluster: {info.get('cluster_name', 'unknown')}")
            print(f"Status: {info.get('status', 'unknown')}")
            return True
        else:
            print(f"Failed to connect: HTTP {response.status_code}")
            return False
    except Exception as e:
        print(f"Connection error: {str(e)}")
        return False


def check_inference_endpoint():
    """Check if ELSER inference endpoint is available, auto-detect if needed."""
    global INFERENCE_ENDPOINT
    
    try:
        response = requests.get(f"{ES_ENDPOINT}/_inference/_all", headers=headers)
        if response.status_code == 200:
            endpoints = response.json().get('endpoints', [])
            
            # First, try to find the specified endpoint
            for endpoint in endpoints:
                if endpoint.get('inference_id') == INFERENCE_ENDPOINT:
                    print(f"Found inference endpoint: {INFERENCE_ENDPOINT}")
                    return True
            
            # If not found, try to auto-detect any ELSER endpoint
            elser_endpoints = []
            for endpoint in endpoints:
                inference_id = endpoint.get('inference_id', '')
                if 'elser' in inference_id.lower():
                    elser_endpoints.append(inference_id)
            
            if elser_endpoints:
                # Prefer endpoints starting with .elser-2- or .elser_model_2
                preferred = [e for e in elser_endpoints if '.elser-2-' in e or '.elser_model_2' in e]
                if preferred:
                    INFERENCE_ENDPOINT = preferred[0]
                else:
                    INFERENCE_ENDPOINT = elser_endpoints[0]
                
                print(f"Specified endpoint not found, using auto-detected: {INFERENCE_ENDPOINT}")
                return True
            
            print(f"Inference endpoint '{INFERENCE_ENDPOINT}' not found")
            print("Available endpoints:")
            for endpoint in endpoints:
                print(f"  - {endpoint.get('inference_id')}")
            return False
        else:
            print(f"Could not check inference endpoints: HTTP {response.status_code}")
            return True  # Continue anyway
    except Exception as e:
        print(f"Error checking inference endpoint: {str(e)}")
        print("Continuing anyway...")
        return True  # Continue anyway


def create_pipeline():
    """Create the PDF processing pipeline."""
    pipeline_path = SCRIPT_DIR.parent / "config" / "pipeline-contracts.json"
    
    try:
        pipeline_config = load_json(pipeline_path)
    except Exception as e:
        print(f"Error loading pipeline config: {e}")
        return False

    try:
        response = requests.put(
            f"{ES_ENDPOINT}/_ingest/pipeline/pdf_pipeline",
            headers=headers,
            json=pipeline_config
        )

        if response.status_code == 200:
            return True
        else:
            print(f"Failed to create pipeline: HTTP {response.status_code}")
            print(json.dumps(response.json(), indent=2))
            return False
    except Exception as e:
        print(f"Error creating pipeline: {str(e)}")
        return False


def check_index_exists():
    """Check if the index already exists."""
    try:
        response = requests.head(f"{ES_ENDPOINT}/{ES_INDEX}", headers=headers)
        return response.status_code == 200
    except:
        return False


def create_index(mapping):
    """Create index with proper mappings."""
    # Delete index if it exists before creating a new one
    if check_index_exists():
        print(f"Deleting existing index '{ES_INDEX}' before import")
        try:
            response = requests.delete(f"{ES_ENDPOINT}/{ES_INDEX}", headers=headers)
            if response.status_code == 200:
                print(f"Index '{ES_INDEX}' deleted")
            else:
                print(f"Failed to delete index: HTTP {response.status_code}")
        except Exception as e:
            print(f"Failed to delete index: {str(e)}")

    # Update mapping with detected inference endpoint
    mapping_with_inference = mapping.copy()
    if 'mappings' in mapping_with_inference and 'properties' in mapping_with_inference['mappings']:
        if 'semantic_content' in mapping_with_inference['mappings']['properties']:
            mapping_with_inference['mappings']['properties']['semantic_content']['inference_id'] = INFERENCE_ENDPOINT

    print(f"Creating index: {ES_INDEX}")
    try:
        response = requests.put(
            f"{ES_ENDPOINT}/{ES_INDEX}",
            headers=headers,
            json=mapping_with_inference
        )

        if response.status_code == 200:
            print(f"Successfully created index: {ES_INDEX}")
            return True
        else:
            print(f"Failed to create index: HTTP {response.status_code}")
            print(json.dumps(response.json(), indent=2))
            return False
    except Exception as e:
        print(f"Error creating index: {str(e)}")
        return False


def get_pdf_files(path):
    """Get list of PDF files from a path."""
    path_obj = Path(path)

    if not path_obj.exists():
        print(f"Path '{path}' does not exist")
        return []

    if path_obj.is_file():
        if path_obj.suffix.lower() == '.pdf':
            return [path_obj]
        else:
            print(f"'{path}' is not a PDF file")
            return []

    elif path_obj.is_dir():
        pdf_files = list(path_obj.glob('*.pdf'))
        if not pdf_files:
            print(f"No PDF files found in directory '{path}'")
        return pdf_files

    return []


def extract_airline_name(filename):
    """Extract airline name from filename."""
    filename_lower = filename.lower()

    # Handle both old and new naming conventions
    if 'american' in filename_lower:
        return 'American Airlines'
    elif 'southwest' in filename_lower:
        return 'Southwest'
    elif 'united' in filename_lower:
        return 'United'
    elif 'delta' in filename_lower or 'dl-' in filename_lower:
        return 'Delta'
    else:
        return 'Unknown'


def index_pdf(pdf_path):
    """Index a single PDF file."""
    pdf_path = Path(pdf_path)
    filename = pdf_path.name
    airline = extract_airline_name(filename)

    try:
        # Read and encode the PDF
        with open(pdf_path, 'rb') as pdf_file:
            encoded_pdf = base64.b64encode(pdf_file.read()).decode('utf-8')

        # Index the document
        index_payload = {
            "data": encoded_pdf,
            "filename": filename,
            "airline": airline
        }

        response = requests.post(
            f"{ES_ENDPOINT}/{ES_INDEX}/_doc?pipeline=pdf_pipeline",
            headers=headers,
            json=index_payload
        )

        if response.status_code in [200, 201]:
            # Don't print here - progress is handled in ingest_pdfs()
            return True
        else:
            # Only print errors, not success messages
            print(f"\nIndexing failed for {filename}: HTTP {response.status_code}")
            print(json.dumps(response.json(), indent=2))
            return False

    except Exception as e:
        print(f"\nError processing {filename}: {str(e)}")
        return False


def ingest_pdfs(pdf_path):
    """Ingest all PDFs from the specified path."""
    pdf_files = get_pdf_files(pdf_path)

    if not pdf_files:
        print("No PDF files to process")
        return False

    total_files = len(pdf_files)
    print(f"Processing {total_files} PDF file(s)...")

    success_count = 0
    failed_count = 0
    processed_count = 0

    for pdf_file in pdf_files:
        if index_pdf(pdf_file):
            success_count += 1
        else:
            failed_count += 1
        
        processed_count += 1
        
        # Update progress
        percentage = round(processed_count / total_files * 100, 1)
        progress = f"\r{processed_count} of {total_files} files processed ({percentage}%)"
        sys.stdout.write(progress)
        sys.stdout.flush()

    # Print newline after progress line
    print()
    
    print(f"Indexed {success_count} of {total_files} file(s)")
    if failed_count > 0:
        print(f"Failed: {failed_count}")

    return failed_count == 0


def verify_ingestion():
    """Verify documents were ingested successfully."""
    try:
        response = requests.get(f"{ES_ENDPOINT}/{ES_INDEX}/_count", headers=headers)
        if response.status_code == 200:
            count = response.json().get('count', 0)
            print(f"Index '{ES_INDEX}' contains {count} document(s)")
            return True
        else:
            print(f"Could not verify document count")
            return True
    except Exception as e:
        print(f"Could not verify document count: {str(e)}")
        return True


def main():
    global ES_ENDPOINT, headers, INFERENCE_ENDPOINT
    
    parser = argparse.ArgumentParser(
        description='Setup Elasticsearch infrastructure and ingest PDF files',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Setup and ingest PDFs from default location
  python3 import_contracts.py

  # Setup and ingest PDFs from specific directory
  python3 import_contracts.py --pdf-path /path/to/pdfs

  # Only setup infrastructure (skip PDF ingestion)
  python3 import_contracts.py --setup-only

  # Skip setup and only ingest PDFs
  python3 import_contracts.py --ingest-only
        """
    )

    parser.add_argument(
        '-c', '--config',
        default=str(DEFAULT_CONFIG_PATH),
        help=f'Path to Elasticsearch config YAML (default: {DEFAULT_CONFIG_PATH})'
    )
    parser.add_argument(
        '-m', '--mapping',
        default=str(DEFAULT_MAPPING_PATH),
        help=f'Path to mappings JSON (default: {DEFAULT_MAPPING_PATH})'
    )
    parser.add_argument(
        '--pdf-path',
        default='data',
        help='Path to PDF file or directory containing PDFs (default: data)'
    )
    parser.add_argument(
        '--setup-only',
        action='store_true',
        help='Only setup infrastructure (pipeline and index), skip PDF ingestion'
    )
    parser.add_argument(
        '--ingest-only',
        action='store_true',
        help='Skip setup, only ingest PDFs (assumes infrastructure exists)'
    )
    parser.add_argument(
        '--inference-endpoint',
        help='Inference endpoint ID (default: .elser-2-elastic, will auto-detect if not found)'
    )

    args = parser.parse_args()
    
    # Set inference endpoint if provided
    if args.inference_endpoint:
        INFERENCE_ENDPOINT = args.inference_endpoint
    
    # Load configuration
    try:
        ES_ENDPOINT, headers = load_config(Path(args.config))
    except Exception as e:
        print(f"Failed to load config: {e}")
        sys.exit(1)

    # Load mapping
    try:
        mapping = load_json(Path(args.mapping).resolve())
    except Exception as e:
        print(f"Failed to load mapping: {e}")
        sys.exit(1)

    # Check Elasticsearch connection
    if not check_elasticsearch():
        print("Cannot connect to Elasticsearch. Exiting.")
        sys.exit(1)

    # Setup phase
    if not args.ingest_only:
        # Check ELSER endpoint
        if not check_inference_endpoint():
            print("ELSER inference endpoint not found!")
            print("Please deploy ELSER via Kibana or API before continuing.")
            print("See: Management → Machine Learning → Trained Models → ELSER → Deploy")
            sys.exit(1)

        # Create pipeline
        if not create_pipeline():
            print("Failed to create pipeline. Exiting.")
            sys.exit(1)

        # Create index (will delete existing one if present)
        if not create_index(mapping):
            print("Failed to create index. Exiting.")
            sys.exit(1)

    # Ingestion phase
    if not args.setup_only:
        import time
        start_time = time.time()

        if not ingest_pdfs(args.pdf_path):
            print("PDF ingestion had errors.")
            sys.exit(1)

        elapsed_time = time.time() - start_time
        print(f"Total ingestion time: {elapsed_time:.2f} seconds")

        # Verify ingestion
        verify_ingestion()


if __name__ == "__main__":
    main()

File: cli-go/import_contracts.go
package main

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/elastic/go-elasticsearch/v8/esapi"
)

const (
	esIndex                = "contracts"
	pipelineName           = "pdf_pipeline"
	defaultInferenceEndpoint = ".elser-2-elastic"
)

// Extend ElasticsearchClient with contract-specific methods
func (c *ElasticsearchClient) CreatePipeline(name string, pipelineConfig map[string]interface{}) error {
	configJSON, err := json.Marshal(pipelineConfig)
	if err != nil {
		return fmt.Errorf("failed to marshal pipeline config: %w", err)
	}

	req := esapi.IngestPutPipelineRequest{
		PipelineID: name,
		Body:       strings.NewReader(string(configJSON)),
	}

	res, err := req.Do(context.Background(), c.client)
	if err != nil {
		return fmt.Errorf("pipeline creation failed: %w", err)
	}
	defer res.Body.Close()

	if res.IsError() {
		body, _ := io.ReadAll(res.Body)
		return fmt.Errorf("pipeline creation failed: %s", string(body))
	}

	c.logger.Printf("Pipeline '%s' created/updated", name)
	return nil
}

func (c *ElasticsearchClient) IndexDocument(indexName string, document map[string]interface{}, pipeline string) error {
	docJSON, err := json.Marshal(document)
	if err != nil {
		return fmt.Errorf("failed to marshal document: %w", err)
	}

	req := esapi.IndexRequest{
		Index:      indexName,
		Body:       strings.NewReader(string(docJSON)),
		Pipeline:   pipeline,
		Refresh:    "wait_for",
	}

	res, err := req.Do(context.Background(), c.client)
	if err != nil {
		return fmt.Errorf("document indexing failed: %w", err)
	}
	defer res.Body.Close()

	if res.IsError() {
		body, _ := io.ReadAll(res.Body)
		return fmt.Errorf("document indexing failed: %s", string(body))
	}

	// Check response body for errors even if status is OK
	var result map[string]interface{}
	if err := json.NewDecoder(res.Body).Decode(&result); err == nil {
		if errVal, ok := result["error"]; ok {
			c.logger.Printf("Elasticsearch returned error in response: %v", errVal)
			return fmt.Errorf("document indexing failed: %v", errVal)
		}
	}

	return nil
}

func (c *ElasticsearchClient) GetInferenceEndpoints() (map[string]interface{}, error) {
	// Use low-level HTTP request since inference endpoints API may not be in typed client
	url := strings.TrimSuffix(c.endpoint, "/") + "/_inference/_all"

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		c.logger.Printf("Failed to create request for inference endpoints: %v", err)
		return map[string]interface{}{"endpoints": []interface{}{}}, nil
	}

	// Add authentication headers from config
	if c.config != nil {
		if c.config.APIKey != "" {
			req.Header.Set("Authorization", "ApiKey "+c.config.APIKey)
		} else if c.config.User != "" && c.config.Password != "" {
			auth := c.config.User + ":" + c.config.Password
			encoded := base64.StdEncoding.EncodeToString([]byte(auth))
			req.Header.Set("Authorization", "Basic "+encoded)
		}

		// Add custom headers
		for k, v := range c.config.Headers {
			req.Header.Set(k, v)
		}
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		c.logger.Printf("Failed to get inference endpoints: %v", err)
		return map[string]interface{}{"endpoints": []interface{}{}}, nil
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		c.logger.Printf("Failed to get inference endpoints: HTTP %d", resp.StatusCode)
		return map[string]interface{}{"endpoints": []interface{}{}}, nil
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		c.logger.Printf("Failed to decode inference endpoints response: %v", err)
		return map[string]interface{}{"endpoints": []interface{}{}}, nil
	}

	return result, nil
}

func (c *ElasticsearchClient) CountDocuments(indexName string) (int64, error) {
	res, err := c.client.Count(c.client.Count.WithIndex(indexName))
	if err != nil {
		c.logger.Printf("Failed to count documents: %v", err)
		return 0, nil
	}
	defer res.Body.Close()

	if res.IsError() {
		c.logger.Printf("Failed to count documents: HTTP %d", res.StatusCode)
		return 0, nil
	}

	var result map[string]interface{}
	if err := json.NewDecoder(res.Body).Decode(&result); err != nil {
		return 0, err
	}

	if count, ok := result["count"].(float64); ok {
		return int64(count), nil
	}

	return 0, nil
}

type ContractLoader struct {
	client           *ElasticsearchClient
	mapping          map[string]interface{}
	inferenceEndpoint string
	logger           *log.Logger
	indexedCount     int
}

func NewContractLoader(client *ElasticsearchClient, mapping map[string]interface{}, logger *log.Logger, inferenceEndpoint string) *ContractLoader {
	if inferenceEndpoint == "" {
		inferenceEndpoint = defaultInferenceEndpoint
	}
	return &ContractLoader{
		client:            client,
		mapping:           mapping,
		inferenceEndpoint: inferenceEndpoint,
		logger:            logger,
		indexedCount:      0,
	}
}

func (l *ContractLoader) CheckElasticsearch() bool {
	health, err := l.client.ClusterHealth()
	if err != nil {
		l.logger.Printf("Connection error: %v", err)
		return false
	}

	clusterName, _ := health["cluster_name"].(string)
	if clusterName == "" {
		clusterName = "unknown"
	}
	status, _ := health["status"].(string)
	if status == "" {
		status = "unknown"
	}

	l.logger.Printf("Cluster: %s", clusterName)
	l.logger.Printf("Status: %s", status)
	return true
}

func (l *ContractLoader) CheckInferenceEndpoint() bool {
	response, err := l.client.GetInferenceEndpoints()
	if err != nil {
		l.logger.Printf("Error checking inference endpoint: %v", err)
		l.logger.Printf("Continuing anyway...")
		return true
	}

	endpoints, _ := response["endpoints"].([]interface{})
	if endpoints == nil {
		endpoints = []interface{}{}
	}

	// First, try to find the specified endpoint
	for _, ep := range endpoints {
		if epMap, ok := ep.(map[string]interface{}); ok {
			if id, ok := epMap["inference_id"].(string); ok && id == l.inferenceEndpoint {
				l.logger.Printf("Found inference endpoint: %s", l.inferenceEndpoint)
				return true
			}
		}
	}

	// Auto-detect ELSER endpoints
	var elserEndpoints []map[string]interface{}
	for _, ep := range endpoints {
		if epMap, ok := ep.(map[string]interface{}); ok {
			if id, ok := epMap["inference_id"].(string); ok {
				if strings.Contains(strings.ToLower(id), "elser") {
					elserEndpoints = append(elserEndpoints, epMap)
				}
			}
		}
	}

	if len(elserEndpoints) > 0 {
		// Prefer endpoints starting with .elser-2- or .elser_model_2
		var preferred []map[string]interface{}
		for _, ep := range elserEndpoints {
			if id, ok := ep["inference_id"].(string); ok {
				if strings.Contains(id, ".elser-2-") || strings.Contains(id, ".elser_model_2") {
					preferred = append(preferred, ep)
				}
			}
		}

		if len(preferred) > 0 {
			if id, ok := preferred[0]["inference_id"].(string); ok {
				l.inferenceEndpoint = id
			}
		} else if len(elserEndpoints) > 0 {
			if id, ok := elserEndpoints[0]["inference_id"].(string); ok {
				l.inferenceEndpoint = id
			}
		}

		l.logger.Printf("Specified endpoint not found, using auto-detected: %s", l.inferenceEndpoint)
		return true
	}

	l.logger.Printf("Inference endpoint '%s' not found", l.inferenceEndpoint)
	l.logger.Printf("Available endpoints:")
	for _, ep := range endpoints {
		if epMap, ok := ep.(map[string]interface{}); ok {
			if id, ok := epMap["inference_id"].(string); ok {
				l.logger.Printf("  - %s", id)
			}
		}
	}
	return false
}

func (l *ContractLoader) CreatePipeline() bool {
	pipelineConfig := map[string]interface{}{
		"description": "Extract text from PDF - semantic_text field handles chunking and embeddings",
		"processors": []map[string]interface{}{
			{
				"attachment": map[string]interface{}{
					"field":         "data",
					"target_field":  "attachment",
					"remove_binary": true,
				},
			},
			{
				"set": map[string]interface{}{
					"field":             "semantic_content",
					"copy_from":         "attachment.content",
					"ignore_empty_value": true,
				},
			},
			{
				"remove": map[string]interface{}{
					"field":          "data",
					"ignore_missing": true,
				},
			},
			{
				"set": map[string]interface{}{
					"field": "upload_date",
					"value": "{{ _ingest.timestamp }}",
				},
			},
		},
	}

	if err := l.client.CreatePipeline(pipelineName, pipelineConfig); err != nil {
		l.logger.Printf("Error creating pipeline: %v", err)
		return false
	}
	return true
}

func (l *ContractLoader) CreateIndex() bool {
	// Delete index if it exists before creating a new one
	exists, err := l.client.IndexExists(esIndex)
	if err != nil {
		l.logger.Printf("Error checking index existence: %v", err)
		return false
	}

	if exists {
		l.logger.Printf("Deleting existing index '%s' before import", esIndex)
		deleted, err := l.client.DeleteIndex(esIndex)
		if err != nil {
			l.logger.Printf("Failed to delete index '%s': %v", esIndex, err)
		} else if deleted {
			l.logger.Printf("Index '%s' deleted", esIndex)
		} else {
			l.logger.Printf("Failed to delete index '%s'", esIndex)
		}
	}

	// Update mapping with detected inference endpoint
	mappingWithInference := deepCopyMap(l.mapping)
	if mappings, ok := mappingWithInference["mappings"].(map[string]interface{}); ok {
		if properties, ok := mappings["properties"].(map[string]interface{}); ok {
			if semanticContent, ok := properties["semantic_content"].(map[string]interface{}); ok {
				semanticContent["inference_id"] = l.inferenceEndpoint
			}
		}
	}

	l.logger.Printf("Creating index: %s", esIndex)
	if err := l.client.CreateIndex(esIndex, mappingWithInference); err != nil {
		l.logger.Printf("Error creating index: %v", err)
		return false
	}
	l.logger.Printf("Successfully created index: %s", esIndex)
	return true
}

func deepCopyMap(m map[string]interface{}) map[string]interface{} {
	jsonData, err := json.Marshal(m)
	if err != nil {
		return m
	}
	var result map[string]interface{}
	if err := json.Unmarshal(jsonData, &result); err != nil {
		return m
	}
	return result
}

func (l *ContractLoader) ExtractAirlineName(filename string) string {
	filenameLower := strings.ToLower(filename)

	if strings.Contains(filenameLower, "american") {
		return "American Airlines"
	} else if strings.Contains(filenameLower, "southwest") {
		return "Southwest"
	} else if strings.Contains(filenameLower, "united") {
		return "United"
	} else if strings.Contains(filenameLower, "delta") || strings.Contains(filenameLower, "dl-") {
		return "Delta"
	}
	return "Unknown"
}

func (l *ContractLoader) GetPdfFiles(path string) ([]string, error) {
	info, err := os.Stat(path)
	if err != nil {
		l.logger.Printf("Path '%s' does not exist", path)
		return []string{}, nil
	}

	if info.IsDir() {
		var pdfFiles []string
		entries, err := os.ReadDir(path)
		if err != nil {
			return nil, err
		}

		for _, entry := range entries {
			if !entry.IsDir() {
				name := entry.Name()
				if strings.HasSuffix(strings.ToLower(name), ".pdf") {
					pdfFiles = append(pdfFiles, filepath.Join(path, name))
				}
			}
		}

		if len(pdfFiles) == 0 {
			l.logger.Printf("No PDF files found in directory '%s'", path)
		}
		return pdfFiles, nil
	}

	// Single file
	if strings.HasSuffix(strings.ToLower(path), ".pdf") {
		return []string{path}, nil
	}

	l.logger.Printf("'%s' is not a PDF file", path)
	return []string{}, nil
}

func (l *ContractLoader) IndexPdf(pdfPath string) bool {
	filename := filepath.Base(pdfPath)
	airline := l.ExtractAirlineName(filename)

	pdfData, err := os.ReadFile(pdfPath)
	if err != nil {
		l.logger.Printf("Error reading %s: %v", filename, err)
		return false
	}

	encodedPdf := base64.StdEncoding.EncodeToString(pdfData)

	document := map[string]interface{}{
		"data":     encodedPdf,
		"filename": filename,
		"airline":  airline,
	}

	if err := l.client.IndexDocument(esIndex, document, pipelineName); err != nil {
		l.logger.Printf("\nError processing %s: %v", filename, err)
		return false
	}

	// Don't log here - progress is handled in IngestPdfs()
	l.indexedCount++
	return true
}

func (l *ContractLoader) IngestPdfs(pdfPath string) bool {
	pdfFiles, err := l.GetPdfFiles(pdfPath)
	if err != nil {
		l.logger.Printf("Error getting PDF files: %v", err)
		return false
	}

	if len(pdfFiles) == 0 {
		l.logger.Printf("No PDF files to process")
		return false
	}

	totalFiles := len(pdfFiles)
	l.logger.Printf("Processing %d PDF file(s)...", totalFiles)

	successCount := 0
	failedCount := 0
	processedCount := 0

	for _, pdfFile := range pdfFiles {
		if l.IndexPdf(pdfFile) {
			successCount++
		} else {
			failedCount++
		}
		
		processedCount++
		
		// Update progress
		percentage := float64(processedCount) / float64(totalFiles) * 100
		fmt.Printf("\r%d of %d files processed (%.1f%%)", processedCount, totalFiles, percentage)
	}

	// Print newline after progress line
	fmt.Println()

	l.logger.Printf("Indexed %d of %d file(s)", successCount, totalFiles)
	if failedCount > 0 {
		l.logger.Printf("Failed: %d", failedCount)
	}

	return failedCount == 0
}

func (l *ContractLoader) VerifyIngestion() {
	time.Sleep(1 * time.Second) // Small delay to ensure documents are searchable

	count, err := l.client.CountDocuments(esIndex)
	if err != nil {
		l.logger.Printf("Could not verify document count: %v", err)
		return
	}

	l.logger.Printf("Index '%s' contains %d document(s)", esIndex, count)

	if count == 0 && l.indexedCount > 0 {
		l.logger.Printf("Warning: Expected %d document(s) but count shows 0. Documents may have failed during pipeline processing.", l.indexedCount)
	}
}

func mainContracts() {
	startTime := time.Now()
	defer func() {
		duration := time.Since(startTime)
		minutes := int(duration.Minutes())
		seconds := duration.Seconds() - float64(minutes*60)
		if minutes > 0 {
			fmt.Printf("\nTotal time: %dm %.2fs\n", minutes, seconds)
		} else {
			fmt.Printf("\nTotal time: %.2fs\n", seconds)
		}
	}()

	var (
		config            = flag.String("c", "config/elasticsearch.yml", "Path to Elasticsearch config YAML")
		configLong        = flag.String("config", "config/elasticsearch.yml", "Path to Elasticsearch config YAML")
		mapping           = flag.String("m", "config/mappings-contracts.json", "Path to mappings JSON")
		mappingLong       = flag.String("mapping", "config/mappings-contracts.json", "Path to mappings JSON")
		pdfPath           = flag.String("pdf-path", "", "Path to PDF file or directory containing PDFs (default: data)")
		setupOnly         = flag.Bool("setup-only", false, "Only setup infrastructure (pipeline and index), skip PDF ingestion")
		ingestOnly        = flag.Bool("ingest-only", false, "Skip setup, only ingest PDFs (assumes infrastructure exists)")
		inferenceEndpoint = flag.String("inference-endpoint", "", "Inference endpoint ID (default: .elser-2-elastic, will auto-detect if not found)")
		status            = flag.Bool("status", false, "Test connection and print cluster health status")
	)

	flag.Parse()

	// Use long form if provided, otherwise use short form
	if *configLong != "config/elasticsearch.yml" {
		config = configLong
	}
	if *mappingLong != "config/mappings-contracts.json" {
		mapping = mappingLong
	}

	logger := log.New(os.Stdout, "", log.LstdFlags)

	esConfig, err := loadConfig(*config)
	if err != nil {
		logger.Fatalf("Error loading config: %v", err)
	}

	client, err := NewElasticsearchClient(esConfig, logger)
	if err != nil {
		logger.Fatalf("Error creating Elasticsearch client: %v", err)
	}

	if *status {
		if err := reportStatus(client, logger); err != nil {
			logger.Fatalf("Error: %v", err)
		}
		return
	}

	mappingData, err := loadMapping(*mapping)
	if err != nil {
		logger.Fatalf("Error loading mapping: %v", err)
	}

	endpoint := *inferenceEndpoint
	if endpoint == "" {
		endpoint = defaultInferenceEndpoint
	}

	loader := NewContractLoader(client, mappingData, logger, endpoint)

	// Check Elasticsearch connection
	if !loader.CheckElasticsearch() {
		logger.Fatalf("Cannot connect to Elasticsearch. Exiting.")
	}

	// Setup phase
	if !*ingestOnly {
		// Check ELSER endpoint
		if !loader.CheckInferenceEndpoint() {
			logger.Fatalf("ELSER inference endpoint not found!")
			logger.Fatalf("Please deploy ELSER via Kibana or API before continuing.")
			logger.Fatalf("See: Management → Machine Learning → Trained Models → ELSER → Deploy")
		}

		// Create pipeline
		if !loader.CreatePipeline() {
			logger.Fatalf("Failed to create pipeline. Exiting.")
		}

		// Create index (will delete existing one if present)
		if !loader.CreateIndex() {
			logger.Fatalf("Failed to create index. Exiting.")
		}
	}

	// Ingestion phase
	if !*setupOnly {
		ingestionStart := time.Now()

		pdfPathValue := *pdfPath
		if pdfPathValue == "" {
			pdfPathValue = resolvePath("data")
		}

		if !loader.IngestPdfs(pdfPathValue) {
			logger.Fatalf("PDF ingestion had errors.")
		}

		elapsed := time.Since(ingestionStart)
		logger.Printf("Total ingestion time: %.2f seconds", elapsed.Seconds())

		// Verify ingestion
		loader.VerifyIngestion()
	}
}
